src/nypl/mod.rs

```diff
@@ -0,0 +1,204 @@
+use crate::Vec2d;
+use std::str::FromStr;
+use custom_error::custom_error;
+use std::sync::Arc;
+use crate::dezoomer::{TilesRect, Dezoomer, DezoomerInput, ZoomLevels, DezoomerError, IntoZoomLevels, DezoomerInputWithContents};
+use std::convert::TryFrom;
+use std::iter::successors;
+use std::fmt::Debug;
+use serde::export::Formatter;
+
+/// A dezoomer for NYPL images
+#[derive(Default)]
+pub struct NYPLImage;
+
+const NYPL_IMAGE_VIEW_PREFIX: &str = "https://digitalcollections.nypl.org/items/";
+const NYPL_META_PREFIX: &str = "https://access.nypl.org/image.php/";
+const NYPL_META_POSTFIX: &str = "/tiles/config.js";
+
+fn _get_image_url_from_meta_url(meta_url: &str) -> String {
+    meta_url.replace(NYPL_META_PREFIX, "")
+        .replace(NYPL_META_POSTFIX, "")
+}
+
+macro_rules! nypl_meta_format_tpl {
+    () => {
+        "https://access.nypl.org/image.php/{}/tiles/config.js"
+    }
+}
+
+impl Dezoomer for NYPLImage {
+    fn name(&self) -> &'static str { "NYPLImage" }
+    fn zoom_levels(&mut self, data: &DezoomerInput) -> Result<ZoomLevels, DezoomerError> {
+        if data.uri.starts_with(NYPL_IMAGE_VIEW_PREFIX) {
+            self.assert(data.uri.contains(NYPL_IMAGE_VIEW_PREFIX))?;
+            let image_id = data.uri.replace(NYPL_IMAGE_VIEW_PREFIX, "");
+            let meta_uri = format!(nypl_meta_format_tpl!(), image_id);
+            Err(DezoomerError::NeedsData { uri: meta_uri })
+        } else {
+            self.assert(data.uri.contains(NYPL_META_PREFIX))?;
+            let DezoomerInputWithContents { uri, contents } = data.with_contents()?;
+            let iter = iter_levels(uri, contents).map_err(DezoomerError::wrap)?;
+            Ok(iter.into_zoom_levels())
+        }
+    }
+}
+
+fn arcs<T>(v: T) -> impl Iterator<Item=Arc<T>> {
+    successors(Some(Arc::new(v)), |x| Some(Arc::clone(x)))
+}
+
+fn iter_levels(uri: &str, contents: &[u8])
+               -> Result<impl Iterator<Item=Level> + 'static, NYPLError> {
+    let base = _get_image_url_from_meta_url(uri);
+    let meta = Metadata::try_from(contents)?;
+    let levels =
+        (0..meta.levels).zip(arcs(base)).zip(arcs(meta))
+            .map(|((level, base), metadata)|
+                Level { metadata, base, level });
+    Ok(levels)
+}
+
+#[derive(PartialEq)]
+struct Level {
+    metadata: Arc<Metadata>,
+    base: Arc<String>,
+    level: u32,
+}
+
+impl Debug for Level {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NYPLImage")
+    }
+}
+
+impl TilesRect for Level {
+    fn size(&self) -> Vec2d {
+        let reverse_level = self.metadata.levels - self.level - 1;
+        self.metadata.size / 2_u32.pow(reverse_level)
+    }
```

Comment:
Did you check how they do the rounding ? Many formats use `ceil(size / 2.^levels)`, or have more complex sizing algorithm, that just happen to match this `floor(size / 2.^levels)` on most but not all image sizes.

Reply:
I haven't done fully test about that.
I've use the code to download an image and it works well, I got the image correctly : )

Reply:
![image](https://user-images.githubusercontent.com/4109722/97863667-4fdf1880-1d42-11eb-95ac-045583934fa7.png)